---
title: "Linear selections in VR and beyond"
author: "Stefan Lang"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette
  
vignette: >
  %\VignetteIndexEntry{cellexalvrR linearSelections}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Linear selections

Linear selections in CellexalVR are simple one group selections. 
The algorithm will only work if the selection has a linear shape.

Internally CellexalvrR clusters the drc data from the selection using kmeans.
These kmean clusters are subsequently provided to slingshot which in turn calculates the
pseudotime 'ranging' from the cluster with the first selected cell to the cluster with the last selected cell. 
Gene expression is afterwards correlated to the calculated pseudotime and the 125 best positively and negatively correlated genes are retrurned to the VR process and displayed as VR heatmap.

## And Beyond

The linear selections turned out to be a very poverful tool in analyzing single cell data as
the structure of the 3D models is best explained with the gradual change in gene expression
in the different areas of the plot. And this linear (pseudotime) analysis is greatliy simplyfied using CellexalVR in comparison to using standalone R.


A linear selection is implemented as cellexalLinear class in cellexalvrR. Every time the user creates a single selection heatmap in VR a cellexalLinear object is created. This object can be accessed from the cellexalObj.RData file in the project output folder after a CellexalVR session has ended.

For this vignett I will use the inbuilt dataset and a saved one group (VR) selection.

```{r,message=FALSE}
library(cellexalvrR)

cellexalObj = reset( cellexalObj )
cellexalObj

```

And the provided one group selection:

```{r,message=FALSE}
head( read.delim( system.file("extdata", "SelectionHSPC_time.txt", package="cellexalvrR"), header=F ) )

```

All linear grouping functions are feeding into the cellexalvrR logging system and therefore we need to start a log before we can efficiently use them:

```{r,message=FALSE}
cellexalObj@outpath = getwd()
cellexalObj = sessionPath( cellexalObj, 'linearExample')
```

Next we will calculate the pseudotime and create the linear correlation.
This part of the analysis would have been run during your VR session.

```{r,message=FALSE}
selectionFile = system.file("extdata", "SelectionHSPC_time.txt",package= "cellexalvrR")
cellexalObj = userGrouping( cellexalObj, selectionFile )
cellexalObj = pseudotimeTest3D(cellexalObj, grouping= cellexalObj@usedObj$lastGroup )
cellexalObj = createStats( "lastEntry" , cellexalObj,  num.sig= 250 )

bossLinearSelection = cellexalObj@usedObj$linearSelections[["lastEntry"]]
cellexalObj = createReport(bossLinearSelection, cellexalObj, info = bossLinearSelection)$cellexalObj


## the statistics table
print ( head( cellexalObj@usedObj$sigGeneLists$lin[[ cellexalObj@usedObj$lastGroup ]] ))

length( cellexalObj@usedObj$deg.genes)
```

You can get information about the existsing linear selections in a cellexalvrR object by printing the linearSelections slot entry.

```{r, message=FALSE}
cellexalObj@usedObj$linearSelections
```

You see the 'lastEntry' linear selection will aloways be in the first place in this list
and the 'lastEntry' is always a copy of one linear selection in the list.

### compareGeneClusters

CompareGeneClusters is a function to compare gene expression profiles between two linearSelections. This functionality is not available in VR.
It requires to have clustered the same genes in two different linear selections.

**How to do this?**

For a real live example the cells in the selection could be separated by treatment, age, desease state or any other research interest.
For this example I can only create an even split between the samples:

```{r, message=FALSE}
linearSelection = getLinearSelection(cellexalObj, bossLinearSelection)

## split the cells in two
subset1 = rownames(linearSelection@dat)[seq(1,nrow(linearSelection@dat),2)]
subset2 = rownames(linearSelection@dat)[seq(2,nrow(linearSelection@dat),2)]

## and split the linear selection in two, too.
linearSubset1 = subset( linearSelection, subset1)
linearSubset2 = subset( linearSelection, subset2)

cellexalObj = addSelection( linearSubset1, cellexalObj )
linearSubset1 = cellexalObj@usedObj$linearSelections[[1]]

cellexalObj = addSelection( linearSubset2, cellexalObj )
linearSubset2 = cellexalObj@usedObj$linearSelections[[1]]

## run the analysis for the two partial linear selections
cellexalObj = createStats( linearSelection, cellexalObj,  num.sig= 250 )
deg.genes = cellexalObj@usedObj$deg.genes

cellexalObj = createReport( linearSubset1, cellexalObj, 
     linearSubset1, deg.genes = deg.genes )$cellexalObj

cellexalObj = createStats( linearSubset2, cellexalObj,  num.sig= 250 )
cellexalObj = createReport(linearSubset2, cellexalObj, 
     linearSubset2, deg.genes = deg.genes )$cellexalObj

## and finally compare them
cellexalObj = compareGeneClusters ( linearSubset1@gname, linearSubset2@gname, 
	 cellexalObj, altGroupNames=c("subset A", "subset B" ) 
)

```

In the end the report has to be finalized:

```{r, message=FALSE}
cellexalObj = renderReport( cellexalObj)
```


## The portable log file

The report is available as PortableLog zip file in the cellexalObj@outpath folder.
[The portable log report created by this vignette](https://github.com/sonejilab/cellexalvrR/blob/master/inst/extdata/PortableLog_linearExample.zip) is also provided with cellexalvrR and can be accessed in R like this:

```
system.file("extdata", "PortableLog_linearExample.zip",package= "cellexalvrR")
```

The file can be unizipped on any computer. Only in the unizipped folder the root html file gives access to all information. All figures and data files are also provided in this folder.


